frontends/web/src/utils/event-legacy.ts

```diff
@@ -0,0 +1,106 @@
+/**
+ * Copyright 2021 Shift Crypto AG
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { apiWebsocket } from '../utils/websocket';
+
+/**
+ * This type describes the subject of an event.
+ */
+
+type Subject = string;
+
+/**
+ * This interface models the legacy events that are received from the backend.
+ */
+
+export interface IEvent {
+    readonly data?: Subject;
+    readonly deviceID?: string;
+    readonly meta?: null;
+    readonly type?: string;
+    readonly code?: string;
+}
+
+/**
+ * This type describes the function used to observe the events.
+ */
+
+ type Observer = (event: IEvent) => void;
+
+/**
+ * This type describes the method returned to unsubscribe again.
+ */
+
+export type IUnsubscribe = () => void;
+
+/**
+ * This interface describes how the subscriptions are stored.
+ */
+
+interface ISubscriptions {
+    [subject: string]: Observer[];
+}
+
+/**
+ * Stores the subscriptions as an object-based hash map.
+ */
+
+const subscriptions: ISubscriptions = {};
+
+/**
+ * This function dispatches the events from the websocket to the observers.
+ */
+
+function handleMessages(data: IEvent): void {
+    if (
+        data.data
+        && typeof data.data === 'string'
+        && data.data in subscriptions
+        && subscriptions[data.data].length
+    ) {
+        // data.data in subscriptions && subscriptions[data.data].length
+        for (const observer of subscriptions[data.data]) {
+            observer(data);
+        }
+    }
+}
+
+/**
+ * Subscribes the given observer on events of the given subject and returns a method to unsubscribe.
+ */
+
+export function subscribe(subject: Subject, observer: Observer): IUnsubscribe {
+    if (!subscriptions[subject]) {
+        subscriptions[subject] = [];
+    }
+    const observers = subscriptions[subject];
+    if (observers.includes(observer)) {
+        console.warn(`observer already registered for ${subject}`);
```

Comment:
feels like this should be a more evident, maybe throw an error?

Reply:
This also comes from the existing non-legacy implementation, IMO this should never happen.

https://github.com/digitalbitbox/bitbox-wallet-app/blob/54ba109819897fb7dd3cf5c827a5f5313506446c/frontends/web/src/utils/event.ts#L98

`console.error` and continue, or `throw `observer already registered for ${subject}`` and end here?

Reply:
I guess it depends. what's the worse case if it happens? if it really can never happen, then why is there a console.warn at all? :)

Reply:
It's just there since it was written 3 years ago

https://github.com/digitalbitbox/bitbox-wallet-app/blame/0260265a9c2e83a833e2ce1b754d5f15fcedfe18/frontends/web/src/utils/event.ts#L92-L94

I don't get why throwing an error what do you want the app to do in such a case.

Reply:
never mind. ok as is.